### PR TITLE
Implement Mac bounce feature

### LINKED ISSUE #514: MacOS: Add a new option to bounce dock icon
Add a new option below `Show red badge on taskbar icon to indicate unread messages`.

By default the setting is disabled. The user can enable the setting, and choose between two options:
 - `once` // only bounces once when a new message is received
 - `until I open the app` // bounces until the app gets focus

image

The icon will always bounce for a new message regardless of the `Show red badge on taskbar icon to indicate unread messages` setting.

### PR DESCRIPTION
**Pull request by wvds**
Closes #514

Before submitting, please confirm you've
 - [x] read and understood our [Contributing Guidelines](https://github.com/mattermost/desktop/blob/master/CONTRIBUTING.md)
 - [x] completed [Mattermost Contributor Agreement](http://www.mattermost.org/mattermost-contributor-agreement/)
 - [x] executed `npm run lint:js` for proper code formatting

Please provide the following information:

**Summary**
With this feature it is be possible to enable the bouncing icon for Mac when receiving a new message. 

**Issue link**
#514

### CODE REVIEW

**Comment by jasonblais** [comment]
This is fantastic! Have asked one of our QA members to help test.

Link to test build: [.../mattermost-desktop-3.7.1-mac.tar.gz](https://1120-43964563-gh.circle-artifacts.com/0/tmp/circle-artifacts.nBoP3rd/mattermost-desktop-3.7.1-mac.tar.gz)

File: src/common/config/defaultPreferences.js [hunk 1]
```diff
@@ -9,7 +9,9 @@ const defaultPreferences = {
   trayIconTheme: 'light',
   minimizeToTray: false,
   notifications: {
-    flashWindow: 0
+    flashWindow: 0,
+    bounceIcon: 0,
```
Comment (yuya-oc) [inline src/common/config/defaultPreferences.js]: The default value should be `false` for boolean preferences.

Comment (wvds) [reply src/common/config/defaultPreferences.js]: Ah indeed, I've changed it!

**Comment by lindalumitchell** [comment]
Hi @yuya-oc and @jasonblais, I spent some time testing this—neat feature! All test cases I tried worked as expected (additional test details below); no issues found! I just had one minor question:

When I first select the setting, neither `once` nor `until I open the app` appears selected by default. I noted that the first time I turned the feature on, the default behavior was `once` (which seems right), so should `once` show as selected by default when the user selects `Bounce the dock icon`?
select-bounce-sub-option-not-selected

Test details, all as expected:
- Bounces as expected (with app behind another app/window or minimized, but not when exited)
- Follows desktop notifications settings (if it would cause a desktop notification it causes a bounce), including not bouncing on an at-mention if desktop notification is set to Never, and including special mentions at-all, at-channel, and at-here
- When set to bounce until I open the app, bounces until I click the app open from dock or from behind another window, or when I click the desktop notification to view the app
- Bounces regardless of red badge setting on the dock

**Comment by jasonblais** [comment]
Nice catch @lindalumitchell! Yes, `once` should be selected by default on the UI. @wvds Can you help take a look?

Once again, big thanks for the PR!

**Comment by wvds** [comment]
@jasonblais Hm, I thought I tested this. Thanks for catching it @lindalumitchell! I'll take a look into it this evening :)

**Comment by wvds** [comment]
@jasonblais @yuya-oc I've updated the pull request with a commit to address this issue. Also fixes a typo which I noticed in another part of the code.

**Comment by jasonblais** [comment]
Thanks! 👍 

@yuya-oc this is ready for your review, and ready to merge if it looks good. We can merge it and include in v4.0 release.

File: src/browser/components/SettingsPage.jsx [hunk 2]
```diff
@@ -419,6 +440,48 @@ const SettingsPage = createReactClass({
...
+        <FormGroup>
+          <Checkbox
+            inline={true}
+            key='bounceIcon'
+            id='inputBounceIcon'
+            ref='bounceIcon'
+            checked={this.state.notifications.bounceIcon}
+            onChange={this.handleBounceIcon}
+            style={{marginRight: '10px'}}
+          >{'Bounce the dock icon'}
```
Comment (yuya-oc) [inline src/browser/components/SettingsPage.jsx]: dock -> Dock

File: src/browser/components/SettingsPage.jsx [hunk 3]
```diff
@@ -419,6 +440,48 @@ const SettingsPage = createReactClass({
...
+          >{'once'}</Radio>
+          {' '}
+          <Radio
+            inline={true}
+            name='bounceIconType'
+            value='critical'
+            disabled={!this.state.notifications.bounceIcon}
+            defaultChecked={this.state.notifications.bounceIconType === 'critical'}
+            onChange={this.handleBounceIconType}
+          >{'until I open the app'}</Radio>
```
Comment (yuya-oc) [inline src/browser/components/SettingsPage.jsx]: @jasonblais Should we capitalize this sentence?

File: src/browser/components/SettingsPage.jsx [hunk 4]
```diff
@@ -419,6 +440,48 @@ const SettingsPage = createReactClass({
...
+            inline={true}
+            name='bounceIconType'
+            value='informational'
+            disabled={!this.state.notifications.bounceIcon}
+            defaultChecked={
+              !this.state.notifications.bounceIconType ||
+              this.state.notifications.bounceIconType === 'informational'
+            }
+            onChange={this.handleBounceIconType}
+          >{'once'}</Radio>
```
Comment (yuya-oc) [inline src/browser/components/SettingsPage.jsx]: @jasonblais Should we capitalize this word?

File: src/browser/components/SettingsPage.jsx [hunk 5]
```diff
@@ -419,6 +440,48 @@ const SettingsPage = createReactClass({
...
+            name='bounceIconType'
+            value='critical'
+            disabled={!this.state.notifications.bounceIcon}
+            defaultChecked={this.state.notifications.bounceIconType === 'critical'}
+            onChange={this.handleBounceIconType}
+          >{'until I open the app'}</Radio>
+          <HelpBlock
+            style={{marginLeft: '20px'}}
+          >
+            {'If enabled, the dock icon bounces once or until the user opens the app when a new message is received.'}
```
Comment (yuya-oc) [inline src/browser/components/SettingsPage.jsx]: dock -> Dock

File: src/common/settings.js [hunk 6]
```diff
@@ -14,7 +14,7 @@ const upgradePreferences = require('./config/upgradePreferences');
 function loadDefault(spellCheckerLocale) {
   const config = JSON.parse(JSON.stringify(defaultPreferences));
   return Object.assign({}, config, {
-    spellCheckerLocale: spellCheckerLocale || defaultPreferences.pellCheckerLocale || 'en-US'
+    spellCheckerLocale: spellCheckerLocale || defaultPreferences.spellCheckerLocale || 'en-US'
```
Comment (yuya-oc) [inline src/common/settings.js]: Possibly this fixes #632 👍

File: src/browser/components/SettingsPage.jsx (hunk 3 above)
Comment (jasonblais) [reply src/browser/components/SettingsPage.jsx]: Now, it'll appear as an option similar to `once`

image

**Comment by jasonblais** [comment]
@yuya-oc @wvds I fixed `dock` --> `Dock` since this was my mistake in the original issue description: https://github.com/mattermost/desktop/issues/514

@yuya-oc For the options, I don't think they need to be capitalized, the design was approved by our UX team and follows the design of other apps.

**Comment by yuya-oc** [comment]
@jason Ah, thanks. I forgot comparing to the picture of the original issue description.

**Pull request merged**